Mark volume as detached when node does not exist for vsphere

FengyunPan opened the pull request:
If node does not exist, node's volumes will be detached
automatically and become available. So mark them detached and
return false without error.
Fix #50266

**Special notes for your reviewer**:
/assign @jingxu97

**Release note**:
```release-note
NONE
```

FengyunPan commented:
/assign @luomiao 
/assign @kerneltime

divyenpatel commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -1073,12 +1070,13 @@ func (vs *VSphere) DisksAreAttached(volPaths []string, nodeName k8stypes.NodeNam
-			return nil, fmt.Errorf("DisksAreAttached failed to determine whether disks %v are still attached: node %q does not exist",
-				volPaths,
-				vSphereInstance)
+			glog.Infof("Node %q does not exist, all the disks on this node will be detached automatically.", nodeName)
```
During host failover when node VM goes to disconnected state or is not accessible for any reason we are returning a Map of volumepath and its attachment status set to false. This will be misinterpreted as disks are already detached from the node and Kubernetes mark volumes as detached after orphaned pod is cleaned up. 

This can causes volumes to remain attached to previous node, and new pod creation on another node will remains in the “containercreating” state. Since both the node vms are powered on and hot volumes can not be attached to other node.

We need to test this PR change for vSphere HA failover.

@BaluDontu @luomiao @tusharnt @rohitjogvmw

luomiao commented:
@FengyunPan Thanks for coming up this PR regarding to the detached volumes.
However, I think this will revert a previous PR to fix another HA issue: https://github.com/kubernetes/kubernetes/pull/45569 
As @divyenpatel has mentioned, directly setting the status of volumes to detached will cause a race condition when the previous dead node is restarted. More explanation can be found in that PR.

Would you please give more information about the reason for this PR?
Do you observe a volume never being detached when a node is removed?
We may need to find a solution to solve both situations.

jingxu97 commented:
@luomiao Could you confirm that when a node is deleted, will the volume be detached automatically? For all other cloud providers such as GCE PD, AWS, that is the behavior. So the function DiskIsAttached() in cloud provider should return false (not attached) without error.

luomiao commented:
@jingxu97 I feel there are two different issues are mixed here :) but the short answer is no, volumes are not detached automatically.

What @FengyunPan 's PR has changed is not when the node is deleted. Instead, it's when the node is **powered off but still exists** in vCenter. In this case, volumes are still attached to the VM. Thus we cannot simply return false (means disk is not attached). The variable and function names are misleading here. Instead of nodeExist, it should be nodeActive.

Regarding to your question, it also depends on how you define "deleting".
1. If "deleting" means "removing the node from kubernetes cluster":
Then the node is till alive in vCenter. So the volume is still attached to the VM.
2. If "deleting" means "removing the node from vCenter":
Removing node from vCenter involves two steps: power off the node and then delete the node.
When the node is powered-off but not deleted yet, the volumes are attached to the VM.

jingxu97 commented:
@luomiao By "deleted" I mean deleted from vCenter, similar to the case of node instance is deleted from cloud provider.

The code nodeExist, err := vs.NodeExists(nodeName) what does it mean? I thought it means the node is deleted from vcenter already?

luomiao commented:
@jingxu97 
The current NodeExists is misleading.
It could return nodeExist=false and err=nil, when the node is powered-off but not deleted.
Now I think it's not a problem with this PR but a problem with the NodeExists function.
Please allow us some time to do a few testings to confirm :)
The change to NodeExists function might be required for this PR.

luomiao commented:
@jingxu97 
Just did some double check. Sorry to take back my words.
NodeExists with a return value nodeExist=false means the node is not in an active status (powered off, disconnected, not found...).
Nodes can be in a non-active state (powered off, disconnected) while volumes are still attached.
In the case of vsphere cloud provider, deleting the node from cloud provider can make the nodes into any of the non-active status (powered off, disconnected, not found...).
But this PR could mark volumes as detached even the volumes might still be attached. Thus it's not correct.

jingxu97 commented:
Is there a way to identify whether the node is deleted from vsphere cloud provider. If no, could we add one?

FengyunPan commented:
>The variable and function names are misleading here. Instead of nodeExist, it should be nodeActive.
NodeExists with a return value nodeExist=false means the node is not in an active status (powered off, disconnected, not found...).
I see, @luomiao thank for your comments.

>Is there a way to identify whether the node is deleted from vsphere cloud provider. If no, could we add one?
I think it is necessary for attachDetachManager, I'd like to check it.

luomiao commented:
@FengyunPan @jingxu97 
Sorry for the delayed reply.
The current PR looks good to me.

We will submit another PR to fix NodeExists (NodeActive) function to return InstanceNotFound error when a node is not existing anymore.

/lgtm

dims commented:
/ok-to-test

jingxu97 commented:
@luomiao should we wait until your fix is merged to merge this PR? Thanks!

divyenpatel commented:
@jingxu97 @luomiao We should wait for the fix, else vSphere HA feature will be broken.

divyenpatel commented:
also We have refactored vSphere Cloud Provider at lot. We can address this PR in the new code base once - https://github.com/kubernetes/kubernetes/pull/49164 is merged.

luomiao commented:
@divyenpatel 
I don't think this PR will break vSphere HA feature though.
This PR has been updated to only return no error when NodeActive function returns err = cloudprovider.InstanceNotFound.
The current NodeActive function doesn't return InstanceNotFound error no matter what state of the node is. Thus it won't change the current code workflow, until we modify the NodeActive function.

Although I agree it's better to hold it after the refactor PR is in :)

FengyunPan commented:
@luomiao @divyenpatel @jingxu97 I have add codes to check if node exists and rebase the PR ( #49164), Please take a look.

FengyunPan commented:
/retest

divyenpatel commented on pkg/cloudprovider/providers/vsphere/vclib/virtualmachine.go:
```diff
@@ -188,9 +188,9 @@ func (vm *VirtualMachine) GetResourcePool(ctx context.Context) (*object.Resource
-func (vm *VirtualMachine) Exists(ctx context.Context) (bool, error) {
+// IsActive checks if the VM is active.
+// Returns true if VM is in poweredOn state.
+func (vm *VirtualMachine) IsActive(ctx context.Context) (bool, error) {
```
Not related to the change made in the PR, but just came across this unwanted code block.
I think this is not required.

```
if vmMoList[0].Summary.Config.Template == false {
 		glog.Warningf("VM is not in %s state", ActivePowerState)
 	} else {
 		glog.Warningf("VM is a template")
}
```

FengyunPan replied:
I think this change of Exists() is helpful for reader, is right?

divyenpatel replied:
Yes `IsActive` is helpful for reader.

divyenpatel commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -584,19 +599,28 @@ func (vs *VSphere) DisksAreAttached(volPaths []string, nodeName k8stypes.NodeNam
 				volPaths,
 				vSphereInstance)
-			return nil, fmt.Errorf("DisksAreAttached failed to determine whether disks %v are still attached: node %q does not exist",
+			return nil, fmt.Errorf("DisksAreAttached failed to determine whether disks %v are still attached: node %q is powered off",
```
Returned error should be different. When the Node VM is powered off, we can still determine whether disks are attached to powered off VM or not. Current error message does not look good in this case.

FengyunPan replied:
Yes, I agree, there is no need to determine wherher node is powered off in DisksAreAttached() and DisksisAttached().

divyenpatel commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -530,15 +540,20 @@ func (vs *VSphere) DiskIsAttached(volPath string, nodeName k8stypes.NodeName) (b
 		}
-		if !nodeExist {
+		if !isActive {
 			glog.Errorf("DiskIsAttached failed to determine whether disk %q is still attached: node %q is powered off",
```
Same as commented for DisksAreAttached. When node VM is powered off, we can still check if disks are attached to the Node VM, and return actual state of the disk.

divyenpatel reviewed (changes requested):
Provided my review comments.

divyenpatel reviewed (approved):
Changes looks good to me.

BaluDontu commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -584,22 +587,19 @@ func (vs *VSphere) DisksAreAttached(volPaths []string, nodeName k8stypes.NodeNam
 		vm, err := vs.getVMByName(ctx, nodeName)
 		if err != nil {
+			if vclib.IsNotFound(err) {
+				glog.Infof("Node %q does not exist, all the disks on this node will be detached automatically.", nodeName)
```
The current log statement is very confusing as when the node is not found, disks will not detached automatically. 
Can you rephrase this info log to
glog.Infof("Node %q does not exist, vsphere CP will assume all disks %v are not attached to it.", nodeName, volPaths).
Same at other places as well.

FengyunPan replied:
Ok, thank for your comment.

Commit:
Mark volume as detached when node does not exist for vsphere

If node does not exist, node's volumes will be detached
automatically and become available. So mark them detached and
return false without error.
Fix #50266

Commit:
[VSphere] Don't return err when node doesn't exist in DetachDisk()

BaluDontu commented on pkg/cloudprovider/providers/vsphere/vsphere.go:
```diff
@@ -584,22 +587,19 @@ func (vs *VSphere) DisksAreAttached(volPaths []string, nodeName k8stypes.NodeNam
 			glog.Errorf("Failed to get VM object for node: %q. err: +%v", vSphereInstance, err)
 			return nil, err
 		}
-		nodeExist, err := vm.Exists(ctx)
```
I have one more question. If node is powered off and the pod is placed on to other available nodes, will the volumes attached to powered off node detach automatically?

What logic in here, would cause it to happen? 

Its because i have one PR which does this https://github.com/kubernetes/kubernetes/pull/40118. Throwing an error when node is powered off caused the volume to be detached automatically.

BaluDontu replied:
Doesn't it matter if the node is powered off or not when the DisksAreAttached() call is made.

FengyunPan replied:
@bacongobbler Have you see the comment from @divyenpatel:"Same as commented for DisksAreAttached. When node VM is powered off, we can still check if disks are attached to the Node VM, and return actual state of the disk."
It seem it doesn't matter if the node is powered off or not when the DisksAreAttached() call is made.
On the other hand, the "vm.Exists(ctx)" just check if vm is power on, can't check if vm is exist.

FengyunPan commented:
@dims @jingxu97 @kerneltime @luomiao @BaluDontu - Any thoughts on it?

BaluDontu commented:
The change looks good to me. /lgtm from my side.

luomiao commented:
sorry for the late reply.
I thought this PR is already merged...
/lgtm
/approve
